storage: add TestWALFailover

jbowens opened the pull request:
Add a new datadriven unit test that tests configuration of WAL failover. Additionally fix a bug where by a reference to a *fs.Env used by WAL failover could be leaked if opening the engine failed.

Epic: none
Release note: none

jbowens committed:
storage: add TestWALFailover

Add a new datadriven unit test that tests configuration of WAL failover.
Additionally fix a bug where by a reference to a *fs.Env used by WAL failover
could be leaked if opening the engine failed.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/121226)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121226#-:-Nu0-jsaA259D0Ki9vx0:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens and @RaduBerinde)

___
*[`pkg/storage/testdata/wal_failover_config` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121226#-Nu0-giT2DzCmdYnufpP:-Nu0-giUDDEk-qGlghWS:b-j6rd9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/50c44e0b427ffa0796b7e2e6ae8cfebb281b516b/pkg/storage/testdata/wal_failover_config#L1)):*
> ```
> mkenv foo
> ```

nice test!


<!-- Sent from Reviewable.io -->

jbowens committed:
storage: require all or none stores to be encrypted for among-stores failover

When WAL failover is enabled, require either all stores to be encrypted or
none. This ensures that data of an encrypted store is never written to an
unencrypted secondary. Add unit tests for encryption-at-rest interactions with
WAL failover.

Epic: none
Release note: none

jbowens reviewed (commented):
TFTR!

Added a second commit with some more testing of encryption-at-rest interactions, and added a new error path when `among-stores` is specified by not all stores are encrypted.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121226#-:-Nu0F2n0EpboMJMsg_eb:b70vhly)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @sumeerbhola)



<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121226#-:-Nu0LN5s-E1sh-ottlSC:b-uwkowd)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens and @RaduBerinde)



<!-- Sent from Reviewable.io -->
